modules/security-jwt/src/main/java/com/devonfw/module/security/jwt/util/TokenExtractor.java

```diff
@@ -0,0 +1,140 @@
+
+	private ObjectMapper objectMapper;
+
+	private KeyStoreAccess keyStoreAccess;
+
+	private Map<String, Object> tokenClaims;
+
+	public TokenExtractor(KeyStoreAccess keyStoreAccess,ObjectMapper objectMapper) {
+		this.keyStoreAccess = keyStoreAccess;
+		this.objectMapper=objectMapper;
+	}
+
+	public Authentication validateTokenAndSignature(String token) {
+
+		Authentication authentication = null;
+		// get algorithm from header
+		// create signatureverifier of particular type
+		// extract token
+		// deserialize it
+		// verify signature
+		// TODO:validate timestamp
+		// TODO: take data from token and create authentication
+
+		String algorithms = JwtHelper.headers(token).get("alg");
+
+		String algorithmFamilyType = getAlgorithmFamilyType(algorithms);
+		if (algorithmFamilyType == null) {
+			// throw exception
+		}
+
+		SignatureVerifier verifier = signatureVerifierFactory(((RSAPublicKey)this.keyStoreAccess.getPublicKey()),
+				algorithmFamilyType);
+
+		Jwt jwt = JwtHelper.decodeAndVerify(token, verifier);
+
+		String claims = jwt.getClaims();
+		try {
+			tokenClaims = this.objectMapper.readValue(claims, Map.class);
+
+			LOG.info("Token Claims " + tokenClaims.toString());
+
+			Instant now = Instant.now();
+
+			if(tokenClaims.get("nbf")!=null) {
+				Instant nbf = Instant.ofEpochSecond(((Number) tokenClaims.get("nbf")).longValue());
+				if (now.isBefore(nbf)) {
+					// Throw exception
+					LOG.info("Token is not valid before " + nbf);
+				}
+
+			}
+
+			if(tokenClaims.get("exp") !=null) {
+				Instant expiration_time = Instant.ofEpochSecond(((Number) tokenClaims.get("exp")).longValue());
+
+				if (now.isAfter(expiration_time)) {
+					// Throw exception
+					LOG.info("Token is expired");
+				}
+			}
+
+
+			String user = tokenClaims.get("sub").toString();
+			// If token do not throw any exception above it is valid and we need to create
+			// authentication object from it
+
+			authentication = new UsernamePasswordAuthenticationToken(user,
+					tokenClaims.get("roles").toString());
+		} catch (IOException e) {
+
+			LOG.error(e.getMessage());
+		}
+		return authentication;
+	}
+
+	private String getAlgorithmFamilyType(String algorithms) {
+		StringBuilder familyAlgorithm = null;
+		List<SignatureAlgorithm> algos = this.keyStoreAccess.loadAllAlgorithmList();
+		for (SignatureAlgorithm algorithm : algos) {
+			if (algorithm.getValue().equalsIgnoreCase(algorithms)) {
+				LOG.info("Algorithm exists--");
+				familyAlgorithm = new StringBuilder(algorithm.getFamilyName());
+				LOG.info("Parent algo is " + familyAlgorithm.toString());
+				return familyAlgorithm.toString();
+			}
+		}
+		return null;
+	}
+
+	public SignatureVerifier signatureVerifierFactory(PublicKey publicKey, String algorithm) {
+		switch (algorithm) {
+		case "Elliptic Curve":
+			return null;
+
+		case "HMAC":
+			return null;
+
+		case "RSA":
+			return new RsaVerifier((RSAPublicKey) publicKey);
+
+		default:
```

Comment:
This seems pretty much incomplete for our `SignatureAlgorithm` enum. Also we should consider making the Enum the factory for `SignatureVerifier` so if you add a new enum constant you are forced to also provide this. Of course this will bundle the Enum with more implementation details and therefore do tight coupling but otherwise we have the same issue here but with the disadvantage the we easily miss a new option here.